### PR TITLE
Social login

### PR DESCRIPTION
**Pull request by bbooth**
*Initial stab at Social Login*
- Added new User Identity domain to hold identity information
- Created InfusionsoftSocialLoginPrincipalFactory which handles the creation of User and UserIdentity records on the fly during authentication.  Attempts to match existing identity by email.
- Override the preconfigured personDirectoryPrincipalResolver for social to use an echoing resolver so it uses our existing user/attributes.

*Not done*
- styling login screen for social providers

### CODE REVIEW

File: cas/src/main/resources/templates/casLoginView.html [hunk 1]
```diff
@@ -1,98 +1,100 @@
...
-            <img src="" th:src="@{${adMobileImageSrcUrl}}" class="iconAdSm img-responsive"/>
-        </a>
-    </div>
-</div>
-
-<div layout:fragment="local_script">
-    <script type="text/javascript" th:src="@{/infusionsoft/js/login.js}"></script>
-</div>
-</body>
+<!DOCTYPE html>
```
Comment (kevin-wise-infusionsoft) [inline cas/src/main/resources/templates/casLoginView.html]: Could you change the line endings back to LF?  Also looks like several other files in your commit are CRLF. Might want to check your github settings so they are always committed as LF.

File: cas/src/main/resources/templates/fragments/loginProviders.html [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+        <ul>
+            <li th:each="entry: ${pac4jUrls}" th:switch="${entry.type}">
+                <a class="btn btn-block btn-social"
+                   th:classappend="'btn-' + ${entry.type}"
+                   th:href="${entry.redirectUrl}"
+                   th:inline="text"
+                   th:title="${entry.name}">
+
+                    <span th:case="twitter" class="fa fa-twitter-square"></span>
```
Comment (kevin-wise-infusionsoft) [inline cas/src/main/resources/templates/fragments/loginProviders.html]: Wow, that's a lot!  Hopefully we'll never use all of those at once!

File: etc/cas/config/cas.properties [hunk 3]
```diff
@@ -4,6 +4,15 @@ cas.server.prefix=${cas.server.name}
 #This is empty to turn off the default CAS users.
 cas.authn.accept.users=
 cas.authn.attributeRepository.defaultAttributesToRelease=id,firstName,lastName,displayName,email,accounts,authorities
+cas.authn.pac4j.facebook.fields=
+cas.authn.pac4j.facebook.id=476094959149593
+cas.authn.pac4j.facebook.secret=b5fc2e5272f252a91c9c744704a6677c
+cas.authn.pac4j.facebook.scope=email
```
Comment (kevin-wise-infusionsoft) [inline etc/cas/config/cas.properties]: What attributes are we getting from Facebook and LinkedIn?  Just email address, or also full name?  If not the full name, how should we prompt the user to complete the profile?  I don't think we'd want to leave those fields blank.  In other words, what does the registration flow look like for social?

File: cas/src/main/java/org/apereo/cas/infusionsoft/authentication/InfusionsoftSocialLoginPrincipalFactory.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            }
+        }
+
+        attributes = userService.createAttributeMapForUser(userIdentity.getUser());
+
+        return super.createPrincipal(userIdentity.getUser().getId().toString(), attributes);
+    }
+
+    private String extractFirstName(Map<String, Object> attributes) {
+        ArrayList<String> possibleKeys = new ArrayList<>();
```
Comment (bbooth) [inline cas/src/main/java/org/apereo/cas/infusionsoft/authentication/InfusionsoftSocialLoginPrincipalFactory.java]: updated to use static.

File: cas/src/main/resources/templates/casLoginView.html (hunk 1 above)
Comment (bbooth) [inline cas/src/main/resources/templates/casLoginView.html]: fixed my gitconfig.

File: etc/cas/config/cas.properties (hunk 3 above)
Comment (bbooth) [inline etc/cas/config/cas.properties]: The default facebook attributes include first name and last name.  See https://developers.facebook.com/docs/facebook-login/permissions/#reference-public_profile

The attributes in the login will be in the canonical form we use now.  So firstName, lastName, displayName,email, etc.

**Pull request merged**